apps/pipelines/nodes/base.py

```diff
@@ -108,10 +108,113 @@ def from_node_output(cls, node_name: str, node_id: str, output: Any = None, **kw
 
         return cls(**kwargs)
 
+    def _get_node_mappings(self):
+        """
+        Helper method to get node ID to name mappings.
+        Returns a tuple of (node_id_to_name, name_to_node_id) dictionaries.
+        """
+        from apps.pipelines.models import Node
+
+        pipeline_version = self.get("pipeline_version")
+        nodes = Node.objects.filter(pipeline__version_number=pipeline_version)
```

Comment:
I'm wondering if the DB calls should not be cached, since they cannot really change during a run. The pipeline state can, but not the DB results

Reply:
What will be the scope of your refactor @snopoke?

I'm feeling a bit iffy about how we currently interact with the pipeline state. Not sure how to exactly express my thoughts or what the correct course of action should be at this point, but I think we should probably get a neat API in place to interact with it sooner rather than later. Anycase, curious to hear your refactoring thoughts

Reply:
We could switch to using a pydantic model for state (https://langchain-ai.github.io/langgraph/how-tos/state-model/#how-to-use-pydantic-model-as-graph-state).

> What will be the scope of your refactor @snopoke?

I want to move all fields out of the state which aren't actually part of the runtime state. Currently this is just `experiment` and `pipeline_version`. But I think it would be useful to provide nodes with a bit more like the DB node object itself and the trace service.